libbeat/reader/debug/debug.go

```diff
@@ -0,0 +1,178 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package debug
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+const (
+	offsetStart        = 100
+	offsetEnd          = 100
+	defaultMinBuffer   = 16 * 1024
+	defaultMaxFailures = 100
+)
+
+type state int
+
+const (
+	initial state = iota
+	running
+	stopped
+)
+
+// CheckFunc func receive a slice of bytes and returns true if it match the predicate.
+type CheckFunc func(offset int64, buf []byte) bool
+
+// Reader is a debug reader used to check the values of specific bytes from an io.Reader,
+// Is is useful is you want to detect if you have received garbage from a network volume.
+type Reader struct {
+	log           *logp.Logger
+	reader        io.Reader
+	buffer        bytes.Buffer
+	minBufferSize int
+	maxFailures   int
+	failures      int
+	predicate     CheckFunc
+	state         state
+	offset        int64
+}
+
+// NewReader returns a debug reader.
+func NewReader(
+	log *logp.Logger,
+	reader io.Reader,
+	minBufferSize int,
+	maxFailures int,
+	predicate CheckFunc,
+) (*Reader, error) {
+	return &Reader{
+		log:           log,
+		minBufferSize: minBufferSize,
+		reader:        reader,
+		maxFailures:   maxFailures,
+		predicate:     predicate,
+	}, nil
+}
+
+// Read will proxy the read call to the original reader and will periodically checks the values of
+// bytes in the buffer.
+func (r *Reader) Read(p []byte) (int, error) {
+	if r.state == stopped {
+		return r.reader.Read(p)
+	}
+
+	if r.state == running && r.failures > r.maxFailures {
+		// cleanup any remaining bytes in the buffer.
+		if r.buffer.Len() > 0 {
+			r.predicate(r.offset, r.buffer.Bytes())
+			r.buffer = bytes.Buffer{}
```

Comment:
Just for safety always run `r.buffer = bytes.Buffer{}`. So changes in how/when buffer is reset has no influence on final cleanup.